test/FunctionsV2/StorageProviderSelectionTests.cs

```diff
@@ -0,0 +1,80 @@
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Azure.WebJobs.Extensions.DurableTask;
+using Microsoft.Azure.WebJobs.Extensions.DurableTask.Tests;
+using Microsoft.Azure.WebJobs.Host.TestCommon;
+using Moq;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace WebJobs.Extensions.DurableTask.Tests.V2
+{
+    public class StorageProviderSelectionTests
+    {
+        static StorageProviderSelectionTests()
+        {
+            AzureStorageDurabilityProviderFactoryMock = new Mock<IDurabilityProviderFactory>();
+            AzureStorageDurabilityProviderFactoryMock.Setup(f => f.Name).Returns("AzureStorage");
+
+            MicrosoftSQLDurabilityProviderFactoryMock = new Mock<IDurabilityProviderFactory>();
+            MicrosoftSQLDurabilityProviderFactoryMock.Setup(f => f.Name).Returns("MicrosoftSQL");
+
+            NetheriteProviderDurabilityProviderFactoryMock = new Mock<IDurabilityProviderFactory>();
+            NetheriteProviderDurabilityProviderFactoryMock.Setup(f => f.Name).Returns("Netherite");
+
+            AllDurabilityProviderFactories = new[] { AzureStorageDurabilityProviderFactoryMock.Object, MicrosoftSQLDurabilityProviderFactoryMock.Object, NetheriteProviderDurabilityProviderFactoryMock.Object };
+            EmptyDurabilityProviderFactoriesList = Enumerable.Empty<IDurabilityProviderFactory>();
+
+            ValidStorageProviderTypesData = new List<object[]>
+            {
+                new object[] { "AzureStorage", AzureStorageDurabilityProviderFactoryMock.Object },
+                new object[] { "MicrosoftSQL", MicrosoftSQLDurabilityProviderFactoryMock.Object },
+                new object[] { "Netherite", NetheriteProviderDurabilityProviderFactoryMock.Object },
+            };
+        }
+
+        public StorageProviderSelectionTests() { }
+
+        private static Mock<IDurabilityProviderFactory> AzureStorageDurabilityProviderFactoryMock { get; }
+
+        private static Mock<IDurabilityProviderFactory> MicrosoftSQLDurabilityProviderFactoryMock { get; }
+
+        private static Mock<IDurabilityProviderFactory> NetheriteProviderDurabilityProviderFactoryMock { get; }
+
+        private static IEnumerable<IDurabilityProviderFactory> AllDurabilityProviderFactories { get; }
+
+        private static IEnumerable<IDurabilityProviderFactory> EmptyDurabilityProviderFactoriesList { get; }
+
+        public static IEnumerable<object[]> ValidStorageProviderTypesData { get; }
+
+        [Theory]
+        [Trait("Category", PlatformSpecificHelpers.TestCategory)]
+        [MemberData(nameof(ValidStorageProviderTypesData))]
+        public void SelectingAvailableStorageProvider(string storageProvider, IDurabilityProviderFactory expectedFactory)
+        {
+            IDurabilityProviderFactory defaultFactory = AllDurabilityProviderFactories.First(f => f.Name.Equals(storageProvider));
```

Comment:
I'm not seeing much value from these unit tests, because it is not actually running through production code, it is instead running against code that is used in production code.

This has two main disadvantages.
1. If the production code changes (and regresses), these tests will still pass, because they are out of sync with actual production code.
2. They don't test end-to-end behavior.

Instead, consider creating an end-to-end run of a JobHost (similar to DurableTaskEndToEndTests) that takes an IEnumerable<IDurabilityProviderFactory> as a parameter fed to it. Feed in the mocks you created above, and then you can call `.Verify()` on those mocks to see which of the IDurabilityProviderFactory objects was used to create the durability provider.

Note, that since you may not be returning an actual Durability provider, you may run into an exception at some point during the `JobHost.Start()` process, but you can just catch that exception gracefully in your test code, and then do the verification.